Added functionality for single rule evaluation

matusmarhefka opened the pull request:
* added '--rule' cmd option for xccdf eval
* usage:
    oscap xccdf eval [--remediate] --profile PROFILE --rule RULE INPUT_FILE

openscap-jenkins commented:
Can one of the admins verify this patch?

mpreisler commented:
@openscap-jenkins add to whitelist

mpreisler commented on utils/oscap-xccdf.c:
```diff
@@ -1041,6 +1046,9 @@ bool getopt_xccdf(int argc, char **argv, struct oscap_action *action)
                 }
+
+		if (action->rule != NULL && action->profile == NULL)
+			return oscap_module_usage(action->module, stderr, "Profile needs to be specified when evaluating a single rule!");
```
Does it really have to? if profile is NULL we just use the default profile so it should all still work, right?

matusmarhefka replied:
Well, I've tried the following on Fedora/RHEL7 (without specifying profile) and it returned zero without evaluating anything:
`$ oscap xccdf eval /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml`
`$ oscap xccdf eval /usr/share/xml/scap/ssg/content/ssg-rhel7-ds.xml`
How exactly can I trigger this "default" profile evaluation?

yuumasato replied:
It doesn't evaluate anything because `default` profile has no rules selected.
I guess the question is whether the one rule cannot be evaluated as if part of the `default` profile, when no profile is provided.

jan-cerny replied:
@matusmarhefka "Default" profile is a virtual profile that is used when no profile is specified. You can trigger it if you run `oscap xccdf eval` without `--profile`.

If no profile is specified, OpenSCAP evaluates all  rules (`<Rule>`) with attribute `selected="true"`, regardless of which profile they belong.

The reason why it does nothing in your case is that in SSG content there is no Rule with attribute `selected="true"`, in other words, the  virtual "Default profile" contains 0 rules.

But in other content, there can be some of there rules. You can have also content with no profiles. For example this file has no profile, but it's a valid XCCDF: https://github.com/OpenSCAP/openscap/blob/maint-1.2/tests/API/XCCDF/unittests/test_xccdf_oval_relative_path.xccdf.xml

matusmarhefka replied:
@jan-cerny Thanks for explanation, I removed this "if" chunk of code and I agree that printing error does not make sense if profile is NULL.

mpreisler commented on utils/oscap-tool.h:
```diff
@@ -118,6 +118,7 @@ struct oscap_action {
 	char *f_verbose_log;
 	/* others */
         char *profile;
+        const char *rule;
```
indents

matusmarhefka replied:
You mean that indents should be tabs? Because this whole structure uses a mix of tabs and spaces for indentation, I was not sure which one to use and AFAIK there is no coding guidelines for openscap project, right?

jan-cerny replied:
@matusmarhefka Yes, it should be tabs. We follow the Linux Kernel coding style. https://www.kernel.org/doc/html/v4.10/process/coding-style.html

Unfortunately, some old code doesn't follow the coding style and uses spaces. In new pull request we use tabs, so that's why there is a mix. We haven't replaced spaces by tabs to avoid merge conflicts and possible problems with downstream patches.

matusmarhefka replied:
Fixed the indent.

shawndwells commented on utils/oscap.8:
```diff
@@ -75,6 +75,11 @@ You may specify OVAL Definition files as the last parameter, XCCDF evaluation wi
+.RS
+Select a particular rule from XCCDF document (also requires profile to be specified). Only this rule will be evaluated.
+.RE
+.TP
```
thank you for putting a note in the man page! This --rule concept has been asked for a lot, and documentation really helps get the word out.

matusmarhefka replied:
@shawndwells Now that you mentioned it, I should probably remove the part which says that profile is required to be specified (based on the discussion above). Thanks.

matusmarhefka commented:
Commit 8734c0e should have all set based on all review comments.

dahaic commented on utils/oscap-xccdf.c:
```diff
@@ -139,6 +139,7 @@ static struct oscap_module XCCDF_EVAL = {
 		"INPUT_FILE - XCCDF file or a source data stream file\n\n"
         "Options:\n"
         "   --profile <name>\r\t\t\t\t - The name of Profile to be evaluated.\n"
+	"   --rule <name>\r\t\t\t\t - The name of a single rule to be evaluated.\n"
```
Indentation should be spaces.

Also can you specify that profile is necessary to specify for this rule, and what are the reasons?

jan-cerny replied:
@dahaic No, we use tabs for indentation. That's a general rule for all our code written in C. As I mentioned in one of previous comments:

>We follow the Linux Kernel coding style. https://www.kernel.org/doc/html/v4.10/process/coding-style.html Unfortunately, some old code doesn't follow the coding style and uses spaces. In new pull request we use tabs, so that's why there is a mix. We haven't replaced spaces by tabs to avoid merge conflicts and possible problems with downstream patches.

Also, profile is not necessary to specify.

dahaic replied:
@jan-cerny please note that lines surrounding this addition are indented by spaces, thus using tab in this particular context is wrong, imho. Either fix it in whole block, or use what is used, and fix it later in separate commit. Sorry for not justifying my objection.

dahaic commented on utils/oscap.8:
```diff
@@ -75,6 +75,11 @@ You may specify OVAL Definition files as the last parameter, XCCDF evaluation wi
 .TP
+\fB\-\-rule RULE\fR
+.RS
+Select a particular rule from XCCDF document. Only this rule will be evaluated.
```
Same as in the --help text:
Can you specify that profile is necessary to be specified for this option and what are the reasons?

mpreisler replied:
`--profile` is not necessary. That's why Matus removed that sentence.

dahaic replied:
@mpreisler wasn't aware {{--profile}} is no longer necessary - still by specifying profile, we get values as specified in that particular profile, right? I am missing this (in my opinion quite useful) behaviour being pointed out.

mpreisler replied:
`(default)` profile is a profile and also can specify Values. As far as the API is concerned it is a proper profile.

I agree that mentioning that --profile affects values that are used with the rules makes sense.

mpreisler commented:
That is not what we do. We have a convention that we fix indentation on the
lines we are changing. We never reindent blocks of code or entire files
because that causes merge conflicts. In this case I don't really care but
imo it's faster to just follow the rule than to bikeshed :-)

On Jun 8, 2017 10:11 AM, "Marek Haičman" <notifications@github.com> wrote:

> *@dahaic* commented on this pull request.
> ------------------------------
>
> In utils/oscap-xccdf.c
> <https://github.com/OpenSCAP/openscap/pull/748#discussion_r120898634>:
>
> > @@ -139,6 +139,7 @@ static struct oscap_module XCCDF_EVAL = {
>  		"INPUT_FILE - XCCDF file or a source data stream file\n\n"
>          "Options:\n"
>          "   --profile <name>\r\t\t\t\t - The name of Profile to be evaluated.\n"
> +	"   --rule <name>\r\t\t\t\t - The name of a single rule to be evaluated.\n"
>
> @jan-cerny <https://github.com/jan-cerny> please note that lines
> surrounding this addition are indented by spaces, thus using tab in this
> particular context is wrong, imho. Either fix it in whole block, or use
> what is used, and fix it later in separate commit. Sorry for not justifying
> my objection.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSCAP/openscap/pull/748#discussion_r120898634>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAt-Af-njlF3DYCPPIUnpq5zh7cI9TWgks5sCAESgaJpZM4Nxdql>
> .
>

matusmarhefka commented:
I added unit tests (make check) and updated the commit message.

Just a note (I also added it into the commit msg): when both profile and rule are specified and the rule is notselected in that profile, it will not be evaluated and 0 will be returned. Do you agree with this behaviour? I based this on the original functionality of 'oscap xccdf eval' without single-rule feature - when some profile is specified and it does not contain any selected rules, oscap returns 0 without evaluating anything.

mpreisler commented on src/XCCDF_POLICY/xccdf_policy.c:
```diff
@@ -950,14 +957,26 @@ _xccdf_policy_rule_evaluate(struct xccdf_policy * policy, const struct xccdf_rul
+	 * mark it as notselected. */
+	if (policy->rule != NULL) {
+		if (strcmp(policy->rule, rule_id) != 0) {
+			dI("Rule '%s' is not selected.", rule_id);
```
This should be dD at best IMO.

matusmarhefka replied:
@mpreisler Good point, it would cause a huge spam. I removed the whole dI line.

mpreisler commented:
I just have one minor nitpick regarding the logging level. Other than that I think this is ready to be merged once jenkins passes.

Thanks for this great contribution!

matusmarhefka commented:
Hopefully the last change, commit eb4e967 changes:
* removed logging about notselected rules
* added a note about values into the man page (thanks to @dahaic)

mpreisler commented:
Fails on Jenkins:
```
+ Test single-rule evaluation                                 FAIL: all.sh
==========================================
1 of 1 test failed
Please report to open-scap-list@redhat.com
==========================================
```

mpreisler commented on tests/API/XCCDF/unittests/Makefile.am:
```diff
@@ -131,6 +131,7 @@ EXTRA_DIST += \
 	test_report_without_oval_poses_no_errors.sh \
 	test_report_without_oval_poses_no_errors.xccdf.xml.result.xml \
 	test_report_without_xsl_fails_gracefully.sh \
+	test_single_rule.sh \
```
Need to add the other resources too, otherwise distcheck is going to fail.

matusmarhefka replied:
Fixed in 4637b13.
